mlir/include/mlir/IR/Properties.td

```diff
@@ -63,6 +65,12 @@ class Property<string storageTypeParam = "", string desc = ""> {
     return convertFromAttribute($_storage, $_attr, $_diag);
   }];
 
+  // The verification predicate for this property. Defaults to And<[]>,
+  // which is trivially true, since properties are always their expected type.
+  // Within the predicate, `$_self` is an instance of the **interface**
+  // type of the property.
+  Pred predicate = ?;
```

Comment:
The trouble with `And<[]>` is that it's hard to detect and I'd like to say "if X doesn't have a predicate, then OptionalProperty<X> doesn't either"

I was considering going an adding explicit True and False cases to the predicate combiner kind but that sameed invasive

... Also, I don't think the And<[]> detector works

Reply:
What is the use case for detecting this/propagating this to `OptionalProperty`? 
Is this purely as an optimization for the generated code (for e.g. debug builds) or do we need this semantics accessible from within TableGen? I'd rather have this optimization implemented in C++ and not have the public API suffer because of it.

Reply:
This is an optimization for the generated code, yeah. I don't want to burden readers or clang with `llvm::all_of(array, [](T&) { true; })` or the like.

I do think there's some value in having the "no predicate" case more visible to the public, since, unlike attributes, most properties will have the trivial predicate, but I'm not married to the idea

Reply:
I thought about it some more and pros- and cons in my head are:
* If you want to optimize the TableGen code such that the code is less (using the various `!if` constructs seen here), both `?` and `And<[]>` semantically are just as capable as they can both be checked for (`!initialized` and `!eq`) and special cased. Admittedly, I think that choosing `And<[]>` as the canonical "true" representation is more arbitrary, but both are to an extent and it is only for "readability of the outputted C++" optimization anyways. It could also be mitigated by defining `defvar True = And<[]>`, making it more readable.
* The con of `?` is that it **requires** classes such as `Optional` to special case `?`, while the `And<[]>` representation would not, making it purely an opt-in optimization, and not a semantic requirement for composing properties. 

Due to the last point I am therefore still leaning towards `And<[]>` as default. I also think in an ideal world it is the backend emitter that would optimize this but that is besides the point.
Would also love to hear other's opinion as I could live with the `?` default if others feel more strongly about it.

Reply:
I've gone and shifted to `And<[]>`, and it looks like `!eq` - at least with a `defvar`, does what I want it to.

It does mean fewer random surprise `?`s than my initial approach